Add support for using a HTTP proxy for outbound connections (just for app install for now)

paulproteus opened the pull request:
This adjusts `installer.js` to use the `request` API, rather than the node
HTTP/HTTPS API. It also adjusts `run-bundle.c++` to pass this variable through,
and adjusts the docs to explain how to use this feature.

Refs #693

Does not fully close #693 yet because a few other places need a similar update.

kentonv commented on shell/server/installer.js:
```diff
@@ -310,92 +310,74 @@ AppInstaller = class AppInstaller {
-    } else if (url.protocol === 'https:') {
+    if (url.protocol === 'http:' || url.protocol === 'https:') {
+      protocol = Request.defaults({
+        maxRedirects: 1,
```
Why only one? Note that we have historically had app URLs that were redirects, and double-redirects are not that uncommon.

paulproteus replied:
How about 5? : D

I just want to pick a constant, any constant.

kentonv replied:
Apparently Chrome and Firefox use a limit of 20 by default.

paulproteus replied:
Fixed in latest push

kentonv commented on shell/server/installer.js:
```diff
@@ -310,92 +310,74 @@ AppInstaller = class AppInstaller {
       throw new Error('Protocol not supported: ' + url.protocol);
     }
 
     // TODO(security):  It could arguably be a security problem that it's possible to probe the
```
Hey I think you're fixing this TODO! (Networks with significant private internals should always specify a proxy.)

paulproteus replied:
Clearly you've never worked at Eventbrite!

Which is to say - I think there should be some other mechanism for avoiding "server side request forgery". And/or we should document the heck out of this.

kentonv commented on src/sandstorm/run-bundle.c++:
```diff
@@ -1162,12 +1162,23 @@ private:
     // locale setting can crash Mongo because we don't have the appropriate locale files available.
+    //
+    // That said, there are a few environment variables that we do re-export.
+    char * http_proxy, * https_proxy;
```
Generally in C++ it's recommended to associate the `*` with the type, i.e. `char* x` not `char *x`. This of course leads to confusion when declaring multiple variables at once as you are doing here, and generally the answer to that is: don't do that. :) That is, declare them on separate lines.

paulproteus replied:
+1

kentonv replied:
Also please use camelCase, not under_scores.

paulproteus replied:
Oh, I was mirroring the environment variable name. +1, I can use camelCase for this.

paulproteus replied:
Fixed by avoiding having so many temp variables in the first place.

kentonv commented on src/sandstorm/run-bundle.c++:
```diff
@@ -1162,12 +1162,23 @@ private:
+    //
+    // That said, there are a few environment variables that we do re-export.
+    char * http_proxy, * https_proxy;
+    KJ_SYSCALL(http_proxy = getenv("https_proxy"));
```
`s/https/http/`

paulproteus replied:
+1

kentonv replied:
`KJ_SYSCALL` only works for functions that return `int`. Also it's only meant for calls that can fail, but getenv() cannot fail (it returns null when there's no match, but that's not a failure).

paulproteus replied:
Cool, I will try without `KJ_SYSCALL`.

paulproteus replied:
Fixed by not using `KJ_SYSCALL` on getenv, merely on setenv, which can error.

kentonv commented on src/sandstorm/run-bundle.c++:
```diff
@@ -1162,12 +1162,23 @@ private:
+    char * http_proxy, * https_proxy;
+    KJ_SYSCALL(http_proxy = getenv("https_proxy"));
+    KJ_SYSCALL(https_proxy = getenv("https_proxy"));
     KJ_SYSCALL(clearenv());
```
I believe that this `clearenv()` call frees the memory that your `http_proxy` and `https_proxy` pointers are pointing at. You will need to copy the strings, preferably using `kj::str()` -- but of course only if they aren't null.

paulproteus replied:
"it seems to work fine" (but I share your assessment from the man page)

+1 to `kj::str()` to copy

paulproteus commented:
@kentonv , how do you suggest I express the following?

```
    const char * keepTheseVars[] = {"http_proxy", "https_proxy"};
    auto numberOfVarsToKeep = 2; // FIXME: How am I supposed to express this & the previous line?
    for (int i = 0; i < numberOfVarsToKeep; i++) {
```

Rationale for doing it this way: I had typos in `https_proxy` vs. `http_proxy` earlier, and so I want to do this in a less repeating-myself fashion to avoid future typos and increase clarity. But I don't know how to make a literal string array in KJ/C++.

Other than the above style question, I believe this addresses all review feedback.

kentonv commented:
`kj::size(keepTheseVars)` will evaluate to `2`.

Also, in C++11 you can write:

    for (char* var: keepTheseVars) {

paulproteus commented:
Thanks @kentonv 

I've pushed a style fix for that, and additionally I've manually re-tested that app installs work, for both http and https app URLs. Ready for your re-review/merge.

paulproteus commented:
Pushed a fix to my "style fix" which makes it actually compile. I will let Jenkins test this; anyway, I did also re-manually-test this locally.

kentonv commented:
I added some OCD style tweaks directly in github. I'll wait for Jenkins to verify I didn't typo anything then merge.
